design/host-profiles.md

```diff
@@ -0,0 +1,2008 @@
+MachineSet using the new profile name, they can bring those hosts into
+the cluster.
+
+### Implementation Details/Notes/Constraints [optional]
+
+It is generally frowned upon to use the Kubernetes API as a "mere
+database". Profiles are less reactive than some other other types of
+resources, but their operator does work in conjunction with the host
+operator to ensure matching is consistently applied. Host configs are
+another example of this pattern.  On the other hand, we might not
+actually need a second controller, if we set the watch rules up in the
+host operator. The choice for how to handle that will be made based on
+which implementation is easier to understand.
+
+### Risks and Mitigations
+
+Cluster admins will have access to APIs to add host profiles on their
+own, leading to deployments on hardware configurations not supported
+by their vendor. The vendor can reject support requests for these
+configurations.
+
+## Design Details
+
+### Work Items
+
+- Define a new BareMetalHostProfile CRD in the baremetal-operator git
+  repo.
+- Add a controller for the BareMetalHostProfile CRD in the
+  baremetal-operator git repo.
+- The logic to match a BareMetalHostProfile (profile) to a
+  BareMetalHost (host) is implemented as a module within the
+  baremetal-operator git repo.
+- A new controller for the profile CRD is added to the
+  baremetal-operator git repo.
+- When a host is created, or a profile triggers its reconciliation,
+  the match logic is invoked by the operator for the host CR, which
+  updates the `metalkube.org/hardware-profile` label on the host to
+  associate it with the profile.
+- Add a tool to the dev-scripts git repo to create virtual machines
+  matching a couple of standard dev/test profiles.
+
+### Dependencies
+
+None
+
+### Test Plan
+
+There will be unit tests for the profile matching logic.
+
+There will be end-to-end tests for the use of profiles to manage
+hosts.
+
+Most developers will run with profiles describing virtual machines,
+which can be varied to ensure that the rest of the project does not
+make assumptions about hardware details.
+
+### Upgrade / Downgrade Strategy
+
+Upgrades of the baremetal-operator can come with additional YAML files
+to define host profiles for common hardware configuration so that they
+are installed automatically when the operator is upgrade.
+
+Existing resources will not require modification on upgrade.
+
+### Version Skew Strategy
+
+This does not apply because the only other consumer of the profile is
+the machine actuator, which only knows the name of a profile and does
+not use any of the details.
+
+## Drawbacks [optional]
+
+See comments in the Risks section above.
+
+## Alternatives [optional]
+
+1. **Build the data into the baremetal-operator.** We would prefer a
+   more data-driven approach that allows us to separate production
+   profiles from dev/test profiles.
+2. **Use a ConfigMap with profile data.**
```

Comment:
I'll look into ConfigMaps more closely to see how to make those work. If we're going to have a config file, maybe that solves the problem of finding an ironic instance when not running inside the cluster, too.

Reply:
One area of concern with using config is that the actuator code will need the same list of names of profiles that the operator receives, so we have to ensure both get exactly the same config file. Is there some mechanism to enforce that consistency?

Reply:
I was thinking the actuator wouldn't need a list of profiles. Something will create a MachineSet that describes a Machine template, including a chosen host profile. Whatever thing creates the MachineSet needs to be able to choose a valid profile. The actuator just needs to match whatever profile it sees specified on a Machine to an available host, without needing to validate it, I think. Does that sound right to you?

My other thought is that it may work well if the authoritative list of profiles, from the perspective of whatever makes the MachineSet, is taken from the list of actual profiles observed on BareMetalHost records. So if it's a UI, it could show the user the list of profiles that have been discovered. Depending on the workflow though of when that decision needs to happen, vs when discovery happens, maybe the list will need to be known in advance. Maybe there's a use case for creating a MachineSet for a hardware profile that hasn't been plugged in yet?

Reply:
I was thinking the machineset or machine would need to validate the profile name, but I guess if the name doesn't match anything then the machine would just never be fulfilled.